src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php

```diff
@@ -102,29 +105,21 @@ public function getEnvironment()
     /**
      * Try to resolve the class name for the widget.
      *
-     * @param PropertyInterface $property The property to get the widget class name for.
+     * @param EnvironmentInterface $environment The environment in use.
+     * @param PropertyInterface    $property    The property to get the widget class name for.
+     * @param ModelInterface|null  $model       The model for which the widget is created.
      *
      * @return string
-     *
-     * @SuppressWarnings(PHPMD.Superglobals)
-     * @SuppressWarnings(PHPMD.CamelCaseVariableName)
      */
-    protected function getWidgetClass(PropertyInterface $property)
+    protected function getWidgetClass(EnvironmentInterface $environment, PropertyInterface $property, $model = null)
     {
-        if (isset($this->widgetMapping[$property->getWidgetType()])) {
-            return $this->widgetMapping[$property->getWidgetType()];
-        }
-
-        if (!isset($GLOBALS['BE_FFL'][$property->getWidgetType()])) {
-            return null;
-        }
+        $dispatcher = $environment->getEventDispatcher();
+        $dispatcher->addListener(DcGeneralEvents::GET_WIDGET_CLASS, [$this, 'retrieveWidgetClass']);
```

Comment:
I don't like this very much.

The event dispatcher must be immutable (at least for Contao 4) and therefore we should rather use this as fallback if the class is still `null` after dispatching.